### PR TITLE
feature/GENESIS-111-tax-rep-date-cust-address

### PR DESCRIPTION
**Pull request by ElaCzapiewska**
answer to global comments:

TaxReportingDate

_- Is it really a field of type date? I understand it as an day or a month when the supplier has to report his taxes every year? Please clarify._

clarified by Vesa - it is a date

_- Should the TaxReportingDate appear alway in the ui - for all countries?
If yes, then we should add at lest the translations for the most customers - like Finnish and German. (I could imagine that at least the German ones will get confused and start asking questions.)_

clarified by Vesa - will only appear for the country where this is mandatory field (as for now CZ only),
still clarification is needed what kind of country is meant (countryofRegistration from BusinessPartner master or country from supplier address field, I will ask Vesa)

InvoiceAddress

_- The global handling and validation on InvoiceAddress duplicate fields doesn't sound right.
Shouldn't we take this as the chance to setup validation on the subcomponent only?_

I removed duplicated fields and validation is now performed directly on invoiceAddress structure,
but I did not take the approach to setup validation on subcomponent only, I left the validation on parent container side.

_- Should the manually added address not being maintainable as long as the SalesInvoice is not sent? (would require a flag in the schema to identify how the InvoiceAddress was filled.)_
I added new bit field to the schema (invoiceAddressIsManual), then invoice address fields are maintainable while the invoice is in draft status and only for manually entered address (where invoiceAddressIsManual is true)

_- What about old but active SalesInvoices without an ReceiverAddress? Now the sender is forced to enter an address. Is this as wanted?_

behavior changed, with the migration script all the existing invoices receive invoiceAddressIsManual=false, then such invoice will not ask for address completion (unless user switches customer -choose it again - then eventually missing address, if still not exists in master data, will be noticed and asked to be completed on invoice)

### CODE REVIEW

File: src/client/components/SalesInvoice/SalesInvoiceConstraints.js [hunk 1]
```diff
@@ -136,6 +136,46 @@ const SalesInvoiceConstraints = {
...
+            tooLong: "^Errors.tooLong"
+        },
+        presence: {
+            message: "^Errors.notNull"
+        }
+    },
+    "country": {
+        presence: {
+            message: "^Errors.notNull"
+        }
```
Comment (nczeranka) [inline src/client/components/SalesInvoice/SalesInvoiceConstraints.js:178]: I don't like the global validation on duplicated fields. But review the global comments and proceed based on the global decision on how to handle the InvoiceAddress.

File: src/client/containers/SalesInvoiceEditor.js [hunk 2]
```diff
@@ -398,8 +415,13 @@ class SalesInvoiceEditor extends Components.ContextComponent
             totalNetPriceWithoutSurchargeNet: this.state.invoiceTotals.totalNetPriceWithoutSurchargeNet,
             items,
             invoiceAddress,
+            street1: invoiceAddress && invoiceAddress.street1,
+            zipCode: invoiceAddress && invoiceAddress.zipCode,
+            city: invoiceAddress && invoiceAddress.city,
+            country: invoiceAddress && invoiceAddress.country,
```
Comment (nczeranka) [inline src/client/containers/SalesInvoiceEditor.js:421]: Why, this duplication? Only for the validation?
Don't like to duplicate those fields. Check the global comment and proceed on decision there.

File: src/client/containers/SalesInvoiceEditor.js [hunk 3]
```diff
@@ -769,7 +792,7 @@ class SalesInvoiceEditor extends Components.ContextComponent
                         customerId: customerId,
                         customer: customer,
                         customerVatId,
-                        invoiceAddress: customerData.invoiceAddress,
+                        invoiceAddress: invoiceAddress || {street1: null, zipCode: null, city: null, country: null, show: true},
```
Comment (nczeranka) [inline src/client/containers/SalesInvoiceEditor.js:795]: see comment above

File: src/client/containers/SalesInvoiceEditor.js [hunk 4]
```diff
@@ -932,6 +955,24 @@ class SalesInvoiceEditor extends Components.ContextComponent
...
+
+    checkShowInvoiceAddressEditor() {
+        const {editMode, customer, invoiceAddress} = this.state;
+
+        return !(editMode === "view") &&
+                customer && customer.id &&
+                (!(invoiceAddress && invoiceAddress.street1 && invoiceAddress.zipCode && invoiceAddress.city && invoiceAddress.country)
+                || (invoiceAddress && invoiceAddress.show));
+    }
+
```
Comment (nczeranka) [inline src/client/containers/SalesInvoiceEditor.js:975]: See comment about "first character of last empty field" in the global comment.

File: src/client/containers/SalesInvoiceEditor.js [hunk 5]
```diff
@@ -1818,12 +1860,34 @@ class SalesInvoiceEditor extends Components.ContextComponent
                                         <label className={this.state.customerValidationError ? "errorLabel" : ""}>
                                             {i18n.getMessage('Labels.invoiceAddress')}
                                         </label>
+                                        {this.checkShowInvoiceAddressEditor() &&
```
Comment (nczeranka) [inline src/client/containers/SalesInvoiceEditor.js:1863]: see above

File: src/client/containers/SalesInvoiceEditor.js [hunk 6]
```diff
@@ -1818,12 +1860,34 @@ class SalesInvoiceEditor extends Components.ContextComponent
...
+                                                <i className="fa fa-info-circle fa-fw" />
+                                            </a>
+                                        }
                                     </div>
                                     <div className='col-md-8'>
                                         {Address.addressToString(this.state.invoiceAddress)}
                                     </div>
                                 </div>
 
+                                {this.checkShowInvoiceAddressEditor() &&
```
Comment (nczeranka) [inline src/client/containers/SalesInvoiceEditor.js:1879]: see above

File: src/server/utils/invoiceDocDefinition.js [hunk 7]
```diff
@@ -236,6 +236,8 @@ module.exports = function (i18n, supplier, customer, invoice, invoiceItems, uoms
         invoiceHeaderItems.push([i18n.getMessage('Labels.bankPaymentReference'), invoice.bankPaymentReference]);
     if (invoice.referenceInformation)
         invoiceHeaderItems.push([i18n.getMessage('Labels.referenceInformation'), invoice.referenceInformation]);
+    if (invoice.taxReportingDate)
+        invoiceHeaderItems.push([i18n.getMessage('Labels.taxReportingDate'), formatISODate(i18n, invoice.taxReportingDate)]);
```
Comment (nczeranka) [inline src/server/utils/invoiceDocDefinition.js:240]: Sounds for me like a sender specific field and I assume we should add it in the Supplier section of the report. Please clarify with Vesa.

**Review by nczeranka — CHANGES_REQUESTED** [review]
TaxReportingDate
* Is it really a field of type date? I understand it as an day or a month when the supplier has to report his taxes every year? Please clarify.
* Should the TaxReportingDate appear alway in the ui - for all countries? 
If yes, then we should add at lest the translations for the most customers - like Finnish and German. (I could imagine that at least the German ones will get confused and start asking questions.)

InvoiceAddress
* The global handling and validation on InvoiceAddress duplicate fields doesn't sound right.
Shouldn't we take this as the chance to setup validation on the subcomponent only?
* Should the manually added address not being maintainable as long as the SalesInvoice is not sent? (would require a flag in the schema to identify how the InvoiceAddress was filled.)
* What about old but active SalesInvoices without an ReceiverAddress? Now the sender is forced to enter an address. Is this as wanted?
Exactly in this case we have an error: On reediting an SalesInvoice with gaps in the InvoiceAddress, the Address editor is opened, but it closes directly on the first char/digit of the last empty field.

File: src/server/utils/invoiceDocDefinition.js (hunk 7 above)
Comment (ElaCzapiewska) [reply src/server/utils/invoiceDocDefinition.js:240]: ok, agreed with Vesa, moved into supplier section

File: src/client/components/SalesInvoice/SalesInvoiceAddress.js [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+                fieldName: 'country',
+                component: (
+                  <this.CountryField
+                    value={invoiceAddress.country}
+                    onChange={e => onChange(e, 'country')}
+                    onBlur = {(e) => forgetField()}
+                    optional={true}
+                  />
+                )
+              })}
```
[the remark this answers is not in the document]
Comment (ElaCzapiewska) [reply src/client/components/SalesInvoice/SalesInvoiceAddress.js:104]: ok, indented

[the remark this answers is not in the document]
Comment (ElaCzapiewska) [reply src/client/components/SalesInvoice/SalesInvoiceAddress.js:104]: and as to:
_- Should the manually added address not being maintainable as long as the SalesInvoice is not sent? (would require a flag in the schema to identify how the InvoiceAddress was filled.)_
I added new bool field to the schema (invoiceAddressIsManual), then invoice address fields are maintainable while the invoice is in draft status and only for manually entered address (where invoiceAddressIsManual is true)

_- What about old but active SalesInvoices without an ReceiverAddress? Now the sender is forced to enter an address. Is this as wanted?_
behavior changed, with the migration script all the existing invoices receive invoiceAddressIsManual=false, then such invoice will not ask for address completion (unless user switches customer -choose it again - then eventually missing address if still not exists in master data will be noticed and asked to be completed on invoice)

File: src/client/components/SalesInvoice/SalesInvoiceConstraints.js (hunk 1 above)
Comment (ElaCzapiewska) [reply src/client/components/SalesInvoice/SalesInvoiceConstraints.js:178]: duplication is removed, validation is performed directly on invoiceAddress fields

File: src/client/containers/SalesInvoiceEditor.js (hunk 2 above)
Comment (ElaCzapiewska) [reply src/client/containers/SalesInvoiceEditor.js:421]: duplication is removed,  only invoiceAddress and its content is used

File: src/client/containers/SalesInvoiceEditor.js (hunk 3 above)
Comment (ElaCzapiewska) [reply src/client/containers/SalesInvoiceEditor.js:795]: this is not duplication, this is just definition of the basic required structure of invoiceAddress if it does not exist in master data

File: src/client/containers/SalesInvoiceEditor.js (hunk 4 above)
Comment (ElaCzapiewska) [reply src/client/containers/SalesInvoiceEditor.js:975]: ok, solved with invoiceAddressIsManual field usage, condition is changed,
behavior also changed - with the migration script all the existing invoices receive invoiceAddressIsManual=false, then such invoice will not show address editor (unless user switches customer -choose it again - then eventually missing address, if still not exists in master data, will be noticed and asked to be completed on invoice)

File: src/client/containers/SalesInvoiceEditor.js (hunk 5 above)
Comment (ElaCzapiewska) [reply src/client/containers/SalesInvoiceEditor.js:1863]: function is changed

File: src/client/containers/SalesInvoiceEditor.js (hunk 6 above)
Comment (ElaCzapiewska) [reply src/client/containers/SalesInvoiceEditor.js:1879]: function is changed

**Review by nczeranka — APPROVED** [review]

**Review by nczeranka — COMMENTED** [review]
Only thing you should check is: How to deal with the TaxReportingDate on CreditNotes?
* Should it be visible for CreditNotes?
* Should it be editable for CreditNotes?

**Pull request merged**